### PR TITLE
Replace deprecated sqlalchemy features that will be removed in sqlalchemy 2.0

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <hkawamura0130@gmail.com>

## Related Issues/PRs

#xxx

## What changes are proposed in this pull request?

[SQLAlchemy 2.0.0b1](https://pypi.org/project/SQLAlchemy/2.0.0b1/) has been published on PyPI. Using this version, this PR replaces deprecated sqlalchemy features that will be removed in 2.0.

https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#migrating-to-sqlalchemy-2-0 is also helpful.

## How is this patch tested?

- [x] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [x] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: .github/workflows/master.yml [hunk 1]
```diff
@@ -144,6 +144,42 @@ jobs:
       run: |
         ./tests/db/compose.sh down --volumes --remove-orphans --rmi all
 
+  ### REMOVE THIS JOS BEFORE MERGE ###
```
Comment (harupy) [inline .github/workflows/master.yml:147]:
```suggestion
  ### REMOVE THIS JOB BEFORE MERGE ###
```

File: mlflow/store/tracking/sqlalchemy_store.py [hunk 2]
```diff
@@ -1370,22 +1373,18 @@ def _get_orderby_clauses(order_by_list, session):
...
-                        (subquery.c.is_nan == sqlalchemy.true(), 1),
-                        (order_value.is_(None), 2),
-                    ],
+                    # Ideally the use of "IS" is preferred here but owing to sqlalchemy
+                    # translation in MSSQL we are forced to use "=" instead.
+                    # These 2 options are functionally identical / unchanged because
+                    # the column (is_nan) is not nullable. However it could become an issue
+                    # if this precondition changes in the future.
+                    (subquery.c.is_nan == sqlalchemy.true(), 1),
+                    (order_value.is_(None), 2),
```
Comment (harupy) [inline mlflow/store/tracking/sqlalchemy_store.py:1382]: https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.case.params.*whens

File: mlflow/store/tracking/sqlalchemy_store.py [hunk 3]
```diff
@@ -163,18 +164,18 @@ def _set_zero_value_insertion_for_autoincrement_column(self, session):
         if self.db_type == MYSQL:
             # config letting MySQL override default
             # to allow 0 value for experiment ID (auto increment column)
-            session.execute("SET @@SESSION.sql_mode='NO_AUTO_VALUE_ON_ZERO';")
+            session.execute(sql.text("SET @@SESSION.sql_mode='NO_AUTO_VALUE_ON_ZERO';"))
```
Comment (harupy) [inline mlflow/store/tracking/sqlalchemy_store.py:167]: https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.text

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 4]
```diff
@@ -2528,15 +2527,13 @@ def _generate_large_data(self, nb_runs=1000):
...
             current_run += 1
-        metrics = pd.DataFrame(metrics_list)
-        metrics.to_sql("metrics", self.store.engine, if_exists="append", index=False)
-        params = pd.DataFrame(params_list)
-        params.to_sql("params", self.store.engine, if_exists="append", index=False)
-        tags = pd.DataFrame(tags_list)
-        tags.to_sql("tags", self.store.engine, if_exists="append", index=False)
-        pd.DataFrame(latest_metrics_list).to_sql(
-            "latest_metrics", self.store.engine, if_exists="append", index=False
-        )
```
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store.py:2539]: Found pandas' SQL features don't work with SQLAlchemy 2.0.

**Comment by harupy** [comment]
https://stackoverflow.com/questions/74089537/sqlalchemy-2-0-sqlbind-parameter-nvarchar-invalid-precision-value-0-microsoft

**Comment by BenWilson2** [comment]
Is the new syntax backwards compatible with versions between 1.3.21 and 2.x?

**Comment by harupy** [comment]
We don't support sqlalchemy < 1.4.0. For 2.x, I ran tests with 2.0.0b1 and they all passed, but this doesn't mean those changes  will be compatible with 2.x because 2.0.0b1 is just a pre-release.

**Comment by BenWilson2** [comment]
> We don't support sqlalchemy < 1.4.0. For 2.x, I ran tests with 2.0.0b1 and they all passed, but this doesn't mean those changes will be compatible with 2.x because 2.0.0b1 is just a pre-release.

Do we need to update https://github.com/mlflow/mlflow/blob/8e7e88da3495e2c287f982d4a7fc20cd51db1139/requirements/test-requirements.txt#L6 ?

**Comment by harupy** [comment]
I think we can remove that line. We specify  the requirement here:

https://github.com/mlflow/mlflow/blob/8e7e88da3495e2c287f982d4a7fc20cd51db1139/requirements/core-requirements.txt#L13

**Review by BenWilson2 — APPROVED** [review]
LGTM!

**Pull request merged**